src/Microsoft.Health.SqlServer/Registration/SqlServerBaseRegistrationExtensions.cs

```diff
@@ -69,14 +69,27 @@ public static IServiceCollection AddSqlServerConnection(this IServiceCollection
 
             services.AddOptions();
             services.TryAddSingleton<ISqlConnectionStringProvider, DefaultSqlConnectionStringProvider>();
-            services.TryAddSingleton<ISqlConnectionFactory>(
+
+            services.TryAddSingleton<ISqlConnectionBuilder>(
+                 p =>
+                 {
+                     var sqlServerDataStoreConfigOption = p.GetRequiredService<IOptions<SqlServerDataStoreConfiguration>>();
+                     SqlServerDataStoreConfiguration config = sqlServerDataStoreConfigOption.Value;
+                     ISqlConnectionStringProvider sqlConnectionStringProvider = p.GetRequiredService<ISqlConnectionStringProvider>();
+                     return config.AuthenticationType == SqlServerAuthenticationType.ManagedIdentity
+                         ? new ManagedIdentitySqlConnectionBuilder(sqlConnectionStringProvider, p.GetRequiredService<IAccessTokenHandler>(), sqlServerDataStoreConfigOption)
+                         : new DefaultSqlConnectionBuilder(sqlConnectionStringProvider, sqlServerDataStoreConfigOption);
+                 });
+
+            // We build SqlConnection object differently for different Authentication Type.
+            // For ManagedIdentity access we set the access token
+            services.TryAddSingleton<ISqlConnectionBuilder>(
```

Comment:
Wouldn't the second`services.TryAddSingleton<ISqlConnectionBuilder>` do nothing if it is already registered above?